File: go/conformance/report_test.go
package conformance

import (
	"encoding/json"
	"strings"
	"testing"
)

func sample() Summary {
	return Summary{Results: []VectorResult{
		{ID: "core/post-executes-once", Tier: "core", Status: "pass", Steps: []StepOutcome{{StepID: "first", Failures: []string{}}}},
		{ID: "core/retry-replays", Tier: "core", Status: "fail", Steps: []StepOutcome{{StepID: "first", Failures: []string{}}, {StepID: "retry", Failures: []string{"status: expected 201, got 409", "body: expected \"<a&b>\""}}}},
		{ID: "core/expiry-executes-again", Tier: "core", Status: "not-applicable", Steps: []StepOutcome{}, Error: "requires short-ttl"},
		{ID: "profile/replayed-header", Tier: "profile", Status: "error", Steps: []StepOutcome{}, Error: "reset returned 500"},
	}, Passed: 1, Failed: 1, NotApplicable: 1, Errored: 1}
}

func TestFormat(t *testing.T) {
	t.Run("REQ-CONF-6: json carries the summary with the TypeScript field names and the target", func(t *testing.T) {
		out, err := Format(sample(), "json", "http://x")
		if err != nil {
			t.Fatal(err)
		}
		var parsed map[string]any
		if err := json.Unmarshal(out, &parsed); err != nil {
			t.Fatal(err)
		}
		if parsed["target"] != "http://x" || parsed["notApplicable"] != float64(1) || parsed["results"].([]any)[1].(map[string]any)["steps"].([]any)[1].(map[string]any)["stepId"] != "retry" {
			t.Fatalf("%v", parsed)
		}
	})
	t.Run("REQ-CONF-6: markdown has the summary line and one row per vector", func(t *testing.T) {
		out, _ := Format(sample(), "markdown", "http://x")
		md := string(out)
		for _, want := range []string{"# anyonce conformance report", "Target: http://x", "1 passed, 1 failed, 1 not applicable, 1 errored", "| Vector | Tier | Status | Details |", "| core/post-executes-once | core | pass |  |", "| core/retry-replays | core | fail | retry: status: expected 201, got 409; retry: body: expected \"<a&b>\" |", "| core/expiry-executes-again | core | not-applicable | requires short-ttl |"} {
			if !strings.Contains(md, want) {
				t.Fatalf("missing %q in\n%s", want, md)
			}
		}
	})
	t.Run("REQ-CONF-6: junit has one testcase per vector with escaped messages", func(t *testing.T) {
		out, _ := Format(sample(), "junit", "")
		xml := string(out)
		for _, want := range []string{`<testsuite name="anyonce-conformance" tests="4" failures="1" errors="1" skipped="1">`, `<testcase classname="core" name="core/post-executes-once"/>`, `<failure message="retry: status: expected 201, got 409; retry: body: expected &quot;&lt;a&amp;b&gt;&quot;"/>`, `<skipped message="requires short-ttl"/>`, `<error message="reset returned 500"/>`} {
			if !strings.Contains(xml, want) {
				t.Fatalf("missing %q in\n%s", want, xml)
			}
		}
		if _, err := Format(sample(), "yaml", ""); err == nil {
			t.Fatal("expected an error for an unknown format")
		}
	})
}

File: packages/conformance/test/report.test.ts
import { describe, expect, test } from 'bun:test';
import { formatReport } from '../src/report';
import type { RunSummary } from '../src/types';

const summary: RunSummary = {
  results: [
    {
      id: 'core/post-executes-once',
      tier: 'core',
      status: 'pass',
      steps: [{ stepId: 'first', failures: [] }],
    },
    {
      id: 'core/retry-replays',
      tier: 'core',
      status: 'fail',
      steps: [
        { stepId: 'first', failures: [] },
        {
          stepId: 'retry',
          failures: ['status: expected 201, got 409', 'handlerInvocations: expected 1, got 2'],
        },
      ],
    },
    {
      id: 'core/expiry-executes-again',
      tier: 'core',
      status: 'not-applicable',
      steps: [],
      error: 'requires short-ttl',
    },
    {
      id: 'profile/replayed-header',
      tier: 'profile',
      status: 'error',
      steps: [],
      error: 'reset returned 500',
    },
  ],
  passed: 1,
  failed: 1,
  notApplicable: 1,
  errored: 1,
};

describe('formatReport', () => {
  test('REQ-CONF-5: json is the summary itself plus the target and the time', () => {
    const parsed = JSON.parse(
      formatReport(summary, 'json', { target: 'http://x', generatedAt: '2026-09-17T00:00:00Z' }),
    );
    expect(parsed.target).toBe('http://x');
    expect(parsed.generatedAt).toBe('2026-09-17T00:00:00Z');
    expect(parsed.passed).toBe(1);
    expect(parsed.results).toHaveLength(4);
  });

  test('REQ-CONF-5: markdown has a summary line and one table row per vector with the failures', () => {
    const md = formatReport(summary, 'markdown', { target: 'http://x' });
    expect(md).toContain('# anyonce conformance report');
    expect(md).toContain('Target: http://x');
    expect(md).toContain('1 passed, 1 failed, 1 not applicable, 1 errored');
    expect(md).toContain('| Vector | Tier | Status | Details |');
    expect(md).toContain('| core/post-executes-once | core | pass |  |');
    expect(md).toContain(
      '| core/retry-replays | core | fail | retry: status: expected 201, got 409; retry: handlerInvocations: expected 1, got 2 |',
    );
    expect(md).toContain(
      '| core/expiry-executes-again | core | not-applicable | requires short-ttl |',
    );
    expect(md).toContain('| profile/replayed-header | profile | error | reset returned 500 |');
  });

  test('REQ-CONF-5: junit has one testcase per vector with failure, skipped and error elements', () => {
    const xml = formatReport(summary, 'junit');
    expect(xml.startsWith('<?xml version="1.0" encoding="UTF-8"?>')).toBe(true);
    expect(xml).toContain(
      '<testsuite name="anyonce-conformance" tests="4" failures="1" errors="1" skipped="1">',
    );
    expect(xml).toContain('<testcase classname="core" name="core/post-executes-once"/>');
    expect(xml).toContain(
      '<testcase classname="core" name="core/retry-replays"><failure message="retry: status: expected 201, got 409; retry: handlerInvocations: expected 1, got 2"/></testcase>',
    );
    expect(xml).toContain(
      '<testcase classname="core" name="core/expiry-executes-again"><skipped message="requires short-ttl"/></testcase>',
    );
    expect(xml).toContain(
      '<testcase classname="profile" name="profile/replayed-header"><error message="reset returned 500"/></testcase>',
    );
  });

  test('REQ-CONF-5: junit escapes XML special characters in messages', () => {
    const xml = formatReport(
      {
        results: [
          {
            id: 'core/x',
            tier: 'core',
            status: 'fail',
            steps: [{ stepId: 's', failures: ['body: expected "<a&b>"'] }],
          },
        ],
        passed: 0,
        failed: 1,
        notApplicable: 0,
        errored: 0,
      },
      'junit',
    );
    expect(xml).toContain('message="s: body: expected &quot;&lt;a&amp;b&gt;&quot;"');
  });
});
